### ISSUE TITLE
Etl load inefficiency ?

### ISSUE THREAD

**Issue by AndreaLanfranchi** [torquem-ch/silkworm#195]
For tracking purposes.

Etl load() method exhibit a great inefficiency when is granted all items are properly sorted across **all** flushed buffers.
This case is what sender's recovery implements: items are collected in the very same order they will be inserted into db.
In this case scenario there is no need to go through a priority queue traversing all flushed buffers to check which one has the lower key on every item being processed.

See https://github.com/torquem-ch/silkworm/blob/1d454b11c9efba66ac81eb3b39c7c9ed67aefc36/db/silkworm/etl/collector.cpp#L102-L157

**Comment by greg7mdp** [comment]
@AndreaLanfranchi not sure how to test the speed, but the following changes may help:
https://github.com/greg7mdp/silkworm/commit/cacbeb7b3db4bbd76f70b8951912519fd31e9055

**Comment by greg7mdp** [comment]
@AndreaLanfranchi I have made another change using `std::map` instead of `std::priority_queue`. I think it may well be faster than using a vector-based priority_queue, but again I don't know how to test it. I have built the code but not tested it.

**Comment by AndreaLanfranchi** [comment]
Hi @greg7mdp,
sorry for late response.
First of all thank you for spending time on the issue. 

I personally think the difference of priority_queue vs map (in this case) is negligible.  At most the container will hold 1 element for each filecollector so the numbers are really low. However a good instrumentation could clear any doubt: unfortunately current Silkworm master branch is off from erigon wrt data paths. I'd suggest to wait till we merge #383 and will eventually show you how to instrument this.

Besides this issue is actually slightly different ... if we know keys are already sorted in every file buffer there is no need for a priority queue nor for a map.

Off topic question : I understand you're using Visual Studio. Did you experience any problem with latest updates ? (I personally can't have the whole project load correctly since update to 16.11.1)

Thank you

**Comment by greg7mdp** [comment]
No problem @AndreaLanfranchi. Yes it would be good to be able to reproduce the issue you mentioned, to validate if any change indeed fixes it.

One issue I corrected in my first checkin was that the comparison function was taking values instead of references. This would cause string copies (and memory allocations) for every comparison, which is inefficient.

> At most the container will hold 1 element for each filecollector so the numbers are really low.

Then the fact that it would be slow is really surprising?

> if we know keys are already sorted in every file buffer there is no need for a priority queue nor for a map

that's true, but you have to make a special case which should not be necessary.

> I understand you're using Visual Studio. Did you experience any problem with latest updates ? (I personally can't have the whole project load correctly since update to 16.11.1).

I use 16.11.0, but I'm not sure it is working as well as it should since it is a first time I'm using cmake in VS and I'm not used to the functionality. It does seem that intellisense is not working right as I often can't navigate to symbol definitions.

**Comment by AndreaLanfranchi** [comment]
> One issue I corrected in my first checkin was that the comparison function was taking values instead of references. This would cause string copies (and memory allocations) for every comparison, which is inefficient.

That't right and we'll be glad to review a PR if you're willing to submit one

> One issue I corrected in my first checkin was that the comparison function was taking values instead of references. This would cause string copies (and memory allocations) for every comparison, which is inefficient.

Same here and eventually on 16.11.1 the CMake configure phase (for VS is Generate Cache) never ends and the IDE hangs.
I just jumped on a trial version of CLion and must say is excellent : just wondering why it took me so much time to evaluate something different from VS. I believe that like many developers grown in the womb of Microsoft I'm lazy enough. ;)

**Comment by greg7mdp** [comment]
Thanks Andrea I just submitted a PR.
I've never tried CLion. I just know the VS debugger so well, and like it so much, that I am reluctant to change. I have heard rumors that the CLion debugger is not as good. Of course the benefit is that it is cross platform.
I'm actually using xemacs as a text editor (even on windows) so it is really the debugger that matters to me most.

**Comment by greg7mdp** [comment]
I saw that Microsoft worked with Kitware to support better cmake integration in VS (see https://devblogs.microsoft.com/cppblog/cmake-presets-integration-in-visual-studio-and-visual-studio-code/). They now support ConfigurePresets instead of ConfigureSettings.json. Also with WSL you can build (and debug) windows and linux binaries on windows from VS.

**Comment by Giulio2002** [comment]
what about replacing Buffer altogheter with std::unique_ptr<Entry>, @AndreaLanfranchi said something about it some time ago if i am not incorrect

**Comment by greg7mdp** [comment]
I Just had a quick look atbuffer.hpp and I don't understand why Buffer::length_ is needed? Why not just rely on buffer_.size()?
std::vector already manages its size and capacity internally.
So the logical thing (imho) would be to call reserve() on the vector for the capacity at creation, and rely in size() to wee how many entries it actually contains.

**Comment by Giulio2002** [comment]
we made tests with @yperbasis and that approach makes the code run approximately 7% faster

**Comment by AndreaLanfranchi** [comment]
@greg7mdp 
 > I Just had a quick look at buffer.hpp and I don't understand why Buffer::length_ is needed? Why not just rely on buffer_.size()?
std::vector already manages its size and capacity internally.

The key answer to your question is [here](https://github.com/torquem-ch/silkworm/blob/master/node/silkworm/etl/buffer.cpp#L23-L34)
You'll notice `buffer_` is never cleared : when `length_==buffer_.size()` a new item is pushed back but in other cases we simply replace the element in buffer at `length_` offset. In this way the vector never reallocates space after a clear.

**Comment by greg7mdp** [comment]
> In this way the vector never reallocates space after a clear.

if you don't want reallocation, you can call `buffer_.resize(0)` instead of `buffer_.clear()`.

> Also we cannot use "reserve" on vector creation as the number of items that will fit is not known: each entry has its own size and the whole buffer is flushed at given threshod of bytes not entries count

hum, this doesn't make sense to me. A `std::vector` is by definition an array of equal size entries. In this case `sizeof(Entry)`, which is a value that is defined and cannot change.
So, while `entry.size()` can vary, because it contains two `std::string` which have an internal pointer, `sizeof(Entry)` is a constant, equal to `2 * sizeof(std::basic_string)`

Also `buffer_` is initialized in the constructor: ` buffer_(kInitialBufferCapacity)`, but that does create a vector of initial size kInitialBufferCapacity, instead of just reserving the capacity.

**Comment by greg7mdp** [comment]
> we made tests with @yperbasis and that approach makes the code run approximately 7% faster

@Giulio2002 I don't think it is likely, most likely you are misusing std::vector. If you share the test with me I'll be happy to update buffer.hpp to make it both fast and streamlined :-)

**Comment by greg7mdp** [comment]
@AndreaLanfranchi I see that `size_` indeed tracks the sum of entry.size(). But I was not referring to the `size_` member.
I was saying that the `length_` member was unnecessary and we could use the vector's size instead.

**Comment by greg7mdp** [comment]
@AndreaLanfranchi I can't right now, but I will suggest a change later in a PR which will maintain the same behavior.

**Comment by Giulio2002** [comment]
> > we made tests with @yperbasis and that approach makes the code run approximately 7% faster
> 
> @Giulio2002 I don't think it is likely, most likely you are misusing std::vector. If you share the test with me I'll be happy to update buffer.hpp to make it both fast and streamlined :-)

Ok, then i will be running transaction lookup on my machine and share the data with @greg7mdp and @AndreaLanfranchi, then @greg7mdp can upload his own modifications on ETL on a fork/branch and will run it and then we see wheter there is any significant change... unfortunately, nor in Erigon, Akula or silkworm we have any benchmarks for stages so we need to run them on nodes to see wheter performance was improved.

**Comment by Giulio2002** [comment]
@greg7mdp @AndreaLanfranchi If you guys agree with this approach, i will run silkworm tx lookup on my Erigon's machine and publish the time that ETL takes.

**Comment by AndreaLanfranchi** [comment]
Of course. I believe we can easily instrument ETL load with a random bytes generator

**Comment by Giulio2002** [comment]
I kind of already do that exact thing in collector_test.cpp, so we can already use the code from there, perhaps after integration we could write a benchmark tool that does these exact things

**Comment by Giulio2002** [comment]
Actually, for some reason my machine has not been able to find any peers for some reason(i am re-running it), so it did not any progress, will report the "bug" to the golang team and will write a benchmark tool at this point...

**Comment by greg7mdp** [comment]
@Giulio2002 I submitted my change in a PR, but it should not change the performance, just make the code simpler. If you have a test demonstrating a performance issue, I'll be happy to take a look at it.

**Comment by Giulio2002** [comment]
@greg7mdp i will be writting one tommorow, I will let you know

**Comment by AndreaLanfranchi** [comment]
I've made a little test which shows the implementation by @greg7mdp effectively shaves some ms from the put() load.
Here some data: the higher the size of dataset the more the original version stays behind.
Implementation [here](https://github.com/torquem-ch/silkworm/blob/buffer/cmd/buffer.cpp)

With a 128MB data set
```
PS C:\Users\Andrea\CMakeBuilds\silkworm\Debug\cmd\Debug> .\buffer.exe

 Feeding buffer base ...
 Done in 21.534s

 [Buffer-Default] First loop ...
 [Buffer-Default] Done in 3.707s
 [Buffer-Default] Second loop ...
 [Buffer-Default] Done in 3.156s


 [Buffer-Gregory] First loop ...
 [Buffer-Gregory] Done in 3.733s
 [Buffer-Gregory] Second loop ...
 [Buffer-Gregory] Done in 3.184s

PS C:\Users\Andrea\CMakeBuilds\silkworm\Debug\cmd\Debug> .\buffer.exe

 Feeding buffer base ...
 Done in 21.514s

 [Buffer-Default] First loop ...
 [Buffer-Default] Done in 3.777s
 [Buffer-Default] Second loop ...
 [Buffer-Default] Done in 3.248s


 [Buffer-Gregory] First loop ...
 [Buffer-Gregory] Done in 3.734s
 [Buffer-Gregory] Second loop ...
 [Buffer-Gregory] Done in 3.249s

PS C:\Users\Andrea\CMakeBuilds\silkworm\Debug\cmd\Debug> .\buffer.exe

 Feeding buffer base ...
 Done in 21.479s

 [Buffer-Default] First loop ...
 [Buffer-Default] Done in 3.568s
 [Buffer-Default] Second loop ...
 [Buffer-Default] Done in 3.015s


 [Buffer-Gregory] First loop ...
 [Buffer-Gregory] Done in 3.529s
 [Buffer-Gregory] Second loop ...
 [Buffer-Gregory] Done in 3.033s
 ```
 
 With a 512MB data set
```
PS C:\Users\Andrea\CMakeBuilds\silkworm\Debug\cmd\Debug> .\buffer.exe

 Feeding buffer base ...
 Done in 1m:25.754s

 [Buffer-Default] First loop ...
 [Buffer-Default] Done in 17.160s
 [Buffer-Default] Second loop ...
 [Buffer-Default] Done in 14.874s


 [Buffer-Gregory] First loop ...
 [Buffer-Gregory] Done in 16.315s
 [Buffer-Gregory] Second loop ...
 [Buffer-Gregory] Done in 14.567s

PS C:\Users\Andrea\CMakeBuilds\silkworm\Debug\cmd\Debug> .\buffer.exe

 Feeding buffer base ...
 Done in 1m:26.337s

 [Buffer-Default] First loop ...
 [Buffer-Default] Done in 16.708s
 [Buffer-Default] Second loop ...
 [Buffer-Default] Done in 14.744s


 [Buffer-Gregory] First loop ...
 [Buffer-Gregory] Done in 16.334s
 [Buffer-Gregory] Second loop ...
 [Buffer-Gregory] Done in 14.571s

PS C:\Users\Andrea\CMakeBuilds\silkworm\Debug\cmd\Debug> .\buffer.exe

 Feeding buffer base ...
 Done in 1m:25.921s

 [Buffer-Default] First loop ...
 [Buffer-Default] Done in 16.050s
 [Buffer-Default] Second loop ...
 [Buffer-Default] Done in 14.411s


 [Buffer-Gregory] First loop ...
 [Buffer-Gregory] Done in 15.705s
 [Buffer-Gregory] Second loop ...
 [Buffer-Gregory] Done in 14.320s
 ```
 
  With a 1GB data set
```
PS C:\Users\Andrea\CMakeBuilds\silkworm\Debug\cmd\Debug> .\buffer.exe

 Feeding buffer base ...
 Done in 2m:52.479s

 [Buffer-Default] First loop ...
 [Buffer-Default] Done in 35.637s
 [Buffer-Default] Second loop ...
 [Buffer-Default] Done in 31.662s


 [Buffer-Gregory] First loop ...
 [Buffer-Gregory] Done in 34.531s
 [Buffer-Gregory] Second loop ...
 [Buffer-Gregory] Done in 30.786s

PS C:\Users\Andrea\CMakeBuilds\silkworm\Debug\cmd\Debug> .\buffer.exe

 Feeding buffer base ...
 Done in 2m:53.264s

 [Buffer-Default] First loop ...
 [Buffer-Default] Done in 35.628s
 [Buffer-Default] Second loop ...
 [Buffer-Default] Done in 32.011s


 [Buffer-Gregory] First loop ...
 [Buffer-Gregory] Done in 35.068s
 [Buffer-Gregory] Second loop ...
 [Buffer-Gregory] Done in 31.455s
 ```
 
 Besides the #388 has less boilerplate code.
 Good catch @greg7mdp

**Comment by AndreaLanfranchi** [comment]
Before you ask the reason of such high timings must warn the above tests were running on `debug` binaries.
The same test for 1GB data set on `release` binaries report
```
PS C:\Users\Andrea\CMakeBuilds\silkworm\Debug\cmd\Debug> ..\..\..\Release\cmd\Release\buffer.exe

 Feeding buffer base ...
 Done in 33.706s

 [Buffer-Default] First loop ...
 [Buffer-Default] Done in 2.470s
 [Buffer-Default] Second loop ...
 [Buffer-Default] Done in 2.713s


 [Buffer-Gregory] First loop ...
 [Buffer-Gregory] Done in 2.491s
 [Buffer-Gregory] Second loop ...
 [Buffer-Gregory] Done in 2.367s

PS C:\Users\Andrea\CMakeBuilds\silkworm\Debug\cmd\Debug> ..\..\..\Release\cmd\Release\buffer.exe

 Feeding buffer base ...
 Done in 33.772s

 [Buffer-Default] First loop ...
 [Buffer-Default] Done in 2.457s
 [Buffer-Default] Second loop ...
 [Buffer-Default] Done in 2.624s


 [Buffer-Gregory] First loop ...
 [Buffer-Gregory] Done in 2.418s
 [Buffer-Gregory] Second loop ...
 [Buffer-Gregory] Done in 2.334s

PS C:\Users\Andrea\CMakeBuilds\silkworm\Debug\cmd\Debug> ..\..\..\Release\cmd\Release\buffer.exe

 Feeding buffer base ...
 Done in 33.740s

 [Buffer-Default] First loop ...
 [Buffer-Default] Done in 2.487s
 [Buffer-Default] Second loop ...
 [Buffer-Default] Done in 2.713s


 [Buffer-Gregory] First loop ...
 [Buffer-Gregory] Done in 2.486s
 [Buffer-Gregory] Second loop ...
 [Buffer-Gregory] Done in 2.362s
```

**Comment by greg7mdp** [comment]
Thanks a lot for the test program @AndreaLanfranchi . I just looked at it and I see that the Buffer API forces an unnecessary copy when inserting the `etl::Entry` into the Buffer. I'll suggest a change that should be significantly faster. Give me a few minutes before merging my PR.

**Comment by greg7mdp** [comment]
Also because the functions are so small I think it makes sense to move them to the header file so they can be inlined. This will make the code faster.

**Comment by greg7mdp** [comment]
I also added the (modified) test that Andrea provided. I removed the sort() so we were testing only the Buffer filling time, and I added a version with the move that destroys the source.

**Comment by Giulio2002** [comment]
Good catch, thank you @greg7mdp

**Comment by greg7mdp** [comment]
> Before you ask the reason of such high timings must warn the above tests were running on debug binaries.

Yes, in my experience it is useless to run performance tests in debug with C++. Whenever I did it the reported bottleneck in the profiler was always incorrect, due to template/STL use and because the lack of function inlining in debug changes the way the code executes significantly. When profiling, I always use a Release build with debug symbols enabled.

**Comment by Giulio2002** [comment]
if there is not anything else to add, shall we close this issue?

**Comment by greg7mdp** [comment]
@Giulio2002 do we have a test case reproducing the slowdown when items are sorted?

I suspect the issue if the vector-based priority_queue. I have a change using `std::map` instead (see https://github.com/greg7mdp/silkworm/commit/64a2a7d9d503eef6b4760f840d0f0847a6bfa02f). This change assumes items added to the collector are unique. If this is not the case I can update it.

Also it is possible that the issue is fixed by my first PR. The comparison function using values instead of references would have been very costly, and the sorted items may have caused a maximum of comparisons.

**Comment by Giulio2002** [comment]
no, but @AndreaLanfranchi said that he noticed this will executing recoveries of senders

**Comment by Giulio2002** [comment]
we can a benchmarks perhaps that does exactly that, instead of random shuffling we just put already sorted keys

**Comment by greg7mdp** [comment]
I think that would be great, @Giulio2002 . If you can provide a test program I'll definitely have a look, otherwise I might be able to write one. I know C++ well, but I just started looking at Silkworm a couple days ago, and I have no experience with eth nodes, so I don't typically understand (yet) what is the purpose of these classes, which is why you can probably write a better test program.

**Comment by Giulio2002** [comment]
oh yeah, will do tommorow morning. also if you have any questions about anything reach out on discord

**Comment by greg7mdp** [comment]
Sounds good, thanks @Giulio2002 ! What's the discord server?

**Comment by Giulio2002** [comment]
here you go: https://discord.gg/Shdhx9Zs

**Comment by Giulio2002** [comment]
#394 made a PR and no real difference was found (at least in etl::Buffer)
